modules/contact/src/userobjects/PenaltyFrictionUserObject.C

```diff
@@ -0,0 +1,351 @@
+
+registerMooseObject("ContactApp", PenaltyFrictionUserObject);
+
+InputParameters
+PenaltyFrictionUserObject::validParams()
+{
+  InputParameters params = WeightedVelocitiesUserObject::validParams();
+  params.addClassDescription(
+      "Computes the mortar frictional contact force via a penalty approach.");
+  params.addRequiredParam<Real>("penalty", "The penalty factor for normal interaction.");
+  params.addParam<Real>("penalty_friction",
+                        "The penalty factor for frictional interaction. If not provide, the normal "
+                        "penalty factor is also used for the frictional problem.");
+  params.addRequiredParam<Real>("friction_coefficient",
+                                "The friction coefficient ruling Coulomb friction equations.");
+  return params;
+}
+
+PenaltyFrictionUserObject::PenaltyFrictionUserObject(const InputParameters & parameters)
+  : WeightedVelocitiesUserObject(parameters),
+    _penalty(getParam<Real>("penalty")),
+    _penalty_friction(isParamValid("penalty_friction") ? getParam<Real>("penalty_friction")
+                                                       : getParam<Real>("penalty")),
+    _friction_coefficient(getParam<Real>("friction_coefficient"))
+{
+  auto check_type = [this](const auto & var, const auto & var_name)
+  {
+    if (!var.isNodal())
+      paramError(var_name,
+                 "The displacement variables must have degrees of freedom exclusively on "
+                 "nodes, e.g. they should probably be of finite element type 'Lagrange'.");
+  };
+  check_type(*_disp_x_var, "disp_x");
+  check_type(*_disp_y_var, "disp_y");
+  if (_has_disp_z)
+    check_type(*_disp_z_var, "disp_z");
+}
+
+const VariableTestValue &
+PenaltyFrictionUserObject::test() const
+{
+  return _disp_x_var->phiLower();
+}
+
+const ADVariableValue &
+PenaltyFrictionUserObject::contactPressure() const
+{
+  return _contact_force;
+}
+
+const ADVariableValue &
+PenaltyFrictionUserObject::contactTangentialPressureDirOne() const
+{
+  return _frictional_contact_force_one;
+}
+
+const ADVariableValue &
+PenaltyFrictionUserObject::contactTangentialPressureDirTwo() const
+{
+  return _frictional_contact_force_two;
+}
+
+void
+PenaltyFrictionUserObject::timestepSetup()
+{
+  WeightedVelocitiesUserObject::timestepSetup();
+
+  _dof_to_old_weighted_gap.clear();
+  _dof_to_old_real_tangential_velocity.clear();
+  _dof_to_old_accumulated_slip.clear();
+  _dof_to_old_normal_pressure.clear();
+
+  for (auto & map_pr : _dof_to_weighted_gap)
+    _dof_to_old_weighted_gap.emplace(map_pr.first, std::move(map_pr.second.first));
+
+  for (auto & map_pr : _dof_to_real_tangential_velocity)
+    _dof_to_old_real_tangential_velocity.emplace(map_pr);
+
+  for (auto & map_pr : _dof_to_normal_pressure)
+    _dof_to_old_normal_pressure.emplace(map_pr.first, MetaPhysicL::raw_value(map_pr.second));
+
+  for (auto & map_pr : _dof_to_accumulated_slip)
+  {
+    if (_dof_to_old_normal_pressure[map_pr.first] >
```

Comment:
Should there always be a key? If so, then I would do `libmesh_map_find` instead of `operator[]`